File: R/kernel.R

# modify to add new kernels ----------------------------------------------------
kVoilaValidKernels = c("exp_kernel", "rq_kernel",
                        "sum_exp_kernels", "exp_const_kernel", "clamped_exp_lin_kernel")

is_valid_kernel_pointer = function(kernel) {
  any(sapply(paste0("Rcpp_", kVoilaValidKernels),
        FUN = inherits, x = kernel))
}

create_kernel_pointer = function(obj, ...) {
  UseMethod("create_kernel_pointer", obj)
}

kVoilaRequiredParameters = list(
  'exp_kernel' = c('amplitude','lengthScales'),
  'rq_kernel' = c('amplitude', 'alpha', 'lengthScales'),
  'sum_exp_kernels' = c('maxAmplitude','amplitude1', 'lengthScales1', 'lengthScales2'),
  'exp_const_kernel' = c('maxAmplitude', 'expAmplitude', 'lengthScales'),
  'clamped_exp_lin_kernel' = c('maxAmplitude', 'linAmplitude', 'linCenter', 'lengthScales')
)

create_kernel_pointer.default = function(type, parameters, inputDimension, epsilon,
                                         lowerBound = NULL, upperBound = NULL) {
  # correct a typical mistake
  if ('scaleLengths' %in% names(parameters)) {
    parameters$lengthScales = parameters$scaleLengths
  }
  kernel = switch(
    type,
    'exp_kernel' = new(exp_kernel, inputDimension, parameters$amplitude,
                       parameters$lengthScales, epsilon),
    'rq_kernel' =  new(rq_kernel, inputDimension, parameters$amplitude,
                       parameters$alpha, parameters$lengthScales, epsilon),
    'sum_exp_kernels' = new(sum_exp_kernels, inputDimension, parameters$maxAmplitude,
                            parameters$amplitude1, parameters$lengthScales1,
                            parameters$lengthScales2, epsilon),
    'exp_const_kernel' = new(exp_const_kernel, inputDimension, parameters$maxAmplitude,
                             parameters$expAmplitude, parameters$lengthScales,
                             epsilon),
    'clamped_exp_lin_kernel' = new(clamped_exp_lin_kernel, inputDimension,
                                   parameters$maxAmplitude,
                                   parameters$linAmplitude, parameters$linCenter,
                                   parameters$lengthScales,
                                   epsilon)
  )
  if (!is.null(lowerBound)) {
    kernel$increase_lower_bound(as.numeric(lowerBound))
  }
  if (!is.null(upperBound)) {
    kernel$decrease_upper_bound(as.numeric(upperBound))
  }
  kernel
}

create_kernel_attributes = function(type, parameters, inputDimension, epsilon) {
  # common to all kernels
  kernelAttributes = list(inputDimension = inputDimension, epsilon = epsilon)

  # the attributes MUST appear in the order used by the cpp constructor
  switch(
    type,
    'exp_kernel' =  {
      kernelAttributes$amplitude = parameters$amplitude
      kernelAttributes$hyperparams = list(lengthScales = parameters$lengthScales)
      kernelAttributes
    },
    'rq_kernel' = {
      kernelAttributes$amplitude = parameters$amplitude
      kernelAttributes$hyperparams = list(alpha = parameters$alpha,
                                          lengthScales = parameters$lengthScales)
      kernelAttributes
    },
    'sum_exp_kernels' = {
      kernelAttributes$maxAmplitude = parameters$maxAmplitude
      kernelAttributes$hyperparams = list(amplitude1 =  parameters$amplitude1,
                                          lengthScales1 = parameters$lengthScales1,
                                          lengthScales2 = parameters$lengthScales2)
      kernelAttributes
    },
    'exp_const_kernel' = {
      kernelAttributes$maxAmplitude = parameters$maxAmplitude
      kernelAttributes$hyperparams = list(expAmplitude = parameters$expAmplitude,
                                          lengthScales = parameters$lengthScales)
      kernelAttributes
    },
    'clamped_exp_lin_kernel' = {
      kernelAttributes$maxAmplitude = parameters$maxAmplitude
      kernelAttributes$hyperparams = list(linAmplitude = parameters$linAmplitude,
                                          linCenter = parameters$linCenter,
                                          lengthScales = parameters$lengthScales)
      kernelAttributes
    })
}

# TODO: ADD ALSO TO THE TYPE ARGUMENT IN THE SDE_KERNEL FUNCTION
# end of required modifications for new kernels --------------------------------

#' Create a gaussian process' kernel
#'
#' Creates a kernel defining the properties of a gaussian process
#' @param type A string specifying the type of kernel to create
#' @param parameters A NAMED list with the parameters that the kernel needs for
#' its proper creation. The easiest way of seeing which parameters are required
#' is to make this function fail. For example:
#' \emph{sde_kernel("exp_kernel",list())}
#' @param inputDimension The input dimension of the kernel
#' @param epsilon A small value to be added to the diagonal of the kernel
#' covariance matrices to regularize them. This improves the numerical stability
#' of the computations.
#' @return A \emph{sde_kernel} S3 object
#' @seealso \code{\link{covmat}}, \code{\link{autocovmat}}, \code{\link{vars}},
#' \code{\link{get_hyperparams}}, \code{\link{set_hyperparams}},
#' \code{\link{decrease_upper_bound}} and \code{\link{increase_lower_bound}}
#' @export
sde_kernel = function(type = c("exp_kernel", "rq_kernel", "sum_exp_kernels",
                               "exp_const_kernel", "clamped_exp_lin_kernel"),
                      parameters, inputDimension = 1, epsilon = 0) {

  type = match.arg(type)
  if ((length(names(parameters)) < length(kVoilaRequiredParameters[[type]])) ||
      !all(names(parameters) %in% kVoilaRequiredParameters[[type]])) {
    stop(paste0("Missing parameters:'", type, "' kernel requires the NAMED parameters c('",
                paste(collapse = "', '", kVoilaRequiredParameters[[type]]), "')"))
  }
  # create kernel just to check the parameters correction (use C++ code to
  # avoid replication) and get default bounds
  junk = create_kernel_pointer(type, parameters, inputDimension, epsilon)

  kernelAttr = create_kernel_attributes(type, parameters, inputDimension, epsilon)
  kernelAttr$lowerBound = lapply(junk$get_lower_bound(), function(x) x)
  kernelAttr$upperBound = lapply(junk$get_upper_bound(), function(x) x)
  names(kernelAttr$upperBound) = names(kernelAttr$lowerBound) = names(kernelAttr$hyperparams)

  attr(kernelAttr, 'type') = type
  class(kernelAttr) = 'sde_kernel'
  kernelAttr
}


create_kernel_pointer.sde_kernel = function(kernel) {
  parameters = kernel
  parameters = c(parameters, kernel$hyperparams)

  create_kernel_pointer.default(attr(kernel, 'type'), parameters,
                                kernel$inputDimension, kernel$epsilon,
                                kernel$lowerBound, kernel$upperBound)
}

#' Covariance matrix
#'
#' Computes the covariance matrix of the input vectors represented with
#' matrices (in which each row represents an input vector).
#'
#' @param kernel An object representing a gaussian process' kernel, e.g. a
#' \emph{sde_kernel} object.
#' @param x,y A matrix in which each row represents an input vector. Note that
#' if the inputs are univariate, a matrix with just one column should
#' be used.
#' @export
covmat = function(kernel, x, y) {
  UseMethod("covmat", kernel)
}


#' @export
covmat.sde_kernel = function(kernel, x, y) {
  kernelPointer = create_kernel_pointer(kernel)
  kernelPointer$covmat(x, y)
}

#' @export
covmat.default = function(kernel, x, y) {
  # use default function to join in a single entry all the kernel pointers
  if (!is_valid_kernel_pointer(kernel)) {
    stop("A C++ kernel pointer was expected")
  }
  kernel$covmat(x, y)
}

#' Autocovariance matrix
#'
#' Computes the autocovariance matrix of the input vectors represented with
#' a matrix (in which each row represents an input vector).
#'
#' @inheritParams covmat
#' @export
autocovmat = function(kernel, x) {
  UseMethod("autocovmat", kernel)
}

#' @export
autocovmat.sde_kernel = function(kernel, x) {
  kernelPointer = create_kernel_pointer(kernel)
  kernelPointer$autocovmat(x)
}

#' @export
autocovmat.default = function(kernel, x) {
  # use default function to join in a single entry all the kernel pointers
  if (!is_valid_kernel_pointer(kernel)) {
    stop("A C++ kernel pointer was expected")
  }
  kernel$autocovmat(x)
}


#' Variance vector
#'
#' Calculates the variances of each of the input vectors represented with
#' a matrix (in which each row represents an input vector).
#'
#' @inheritParams autocovmat
#' @export
vars = function(kernel, x) {
  UseMethod("vars", kernel)
}

#' @export
vars.sde_kernel = function(kernel, x) {
  kernelPointer = create_kernel_pointer(kernel)
  kernelPointer$vars(x)
}

#' @export
vars.default = function(kernel, x) {
  # use default function to join in a single entry all the kernel pointers
  if (!is_valid_kernel_pointer(kernel)) {
    stop("A C++ kernel pointer was expected")
  }
  kernel$vars(x)
}

#' Get the hyperparameters of a kernel
#' @inheritParams autocovmat
#' @export
get_hyperparams = function(kernel) {
  UseMethod("get_hyperparams", kernel)
}

#' @export
get_hyperparams.sde_kernel = function(kernel) {
  kernel$hyperparams
}

#' Set the hyperparameters of a kernel
#' @inheritParams autocovmat
#' @param hyperparams The new hyperparameters of the kernel
#' @export
set_hyperparams = function(kernel, hyperparams) {
  UseMethod("set_hyperparams", kernel)
}

#' @export
set_hyperparams.sde_kernel = function(kernel, hyperparams) {
  if (is.list(hyperparams)) {
    hyperparams = arrange_hyperparams_list(kernel, hyperparams)
  } else {
    # hyperparams is a vector: we assume default order.
    # get name and sizes of the hyperparams
    oldHp = get_hyperparams(kernel)
    hpNames = names(oldHp)
    hpDims = sapply(oldHp, length)
    begin = c(1, head(cumsum(hpDims), -1) + 1)
    end = begin + hpDims - 1
    hyperparamsList = list()
    for (i in seq_along(hpNames)) {
     hyperparamsList[[ hpNames[i] ]] = hyperparams[ begin[i]:end[i] ]
    }
    hyperparams = hyperparamsList
  }
  # avoid replication of code using the hyperparameters' checks included in
  # C++ code
  kernelPointer = create_kernel_pointer(kernel)
  kernelPointer$set_hyperparams(as.numeric(hyperparams))
  # if no exception is given the parameters are correct...
  kernel$hyperparams = hyperparams
  kernel
}

#' @export
set_hyperparams.default = function(kernel, hyperparams) {
  # use default function to join in a single entry all the kernel pointers
  if (!is_valid_kernel_pointer(kernel)) {
    stop("A C++ kernel pointer was expected")
  }
  kernel$set_hyperparams(as.numeric(hyperparams))
  kernel
}


#' Decreases the upper bound of the kernel
#'
#' Decreases the maximum values that each of the hyperparameters of the kernel
#' may take.
#'
#' @inheritParams autocovmat
#' @param upperBound A vector with the new values of the upper bound.
#' @export
decrease_upper_bound = function(kernel, upperBound) {
  UseMethod("decrease_upper_bound", kernel)
}

#' @export
decrease_upper_bound.sde_kernel = function(kernel, upperBound) {
  upperBound = arrange_hyperparams_list(kernel, upperBound)
  # avoid replication of code using the checks included in C++ code
  kernelPointer = create_kernel_pointer(kernel)
  kernelPointer$decrease_upper_bound(as.numeric(upperBound))
  # if no exception is given the parameters are correct...
  kernel$upperBound = upperBound
  kernel
}

#' Decreases the lower bound of the kernel
#'
#' Decreases the minimum values that each of the hyperparameters of the kernel
#' may take.
#'
#' @inheritParams autocovmat
#' @param lowerBound A vector with the new values of the lower bound.
#' @export
increase_lower_bound = function(kernel, lowerBound) {
  UseMethod("increase_lower_bound", kernel)
}

#' @export
increase_lower_bound = function(kernel, lowerBound) {
  lowerBound = arrange_hyperparams_list(kernel, lowerBound)
  # avoid replication of code using the checks included in C++ code
  kernelPointer = create_kernel_pointer(kernel)
  kernelPointer$increase_lower_bound(as.numeric(lowerBound))
  # if no exception is given the parameters are correct...
  kernel$lowerBound = lowerBound
  kernel
}


arrange_hyperparams_list = function(kernel, hyperparams) {
    if ((length(names(hyperparams)) < length(kernel$hyperparams)) ||
      !all(names(hyperparams) %in% names(kernel$hyperparams))) {
    stop(paste0("Missing hyperparameter in  'hyperparams'. ",
                "The following NAMED hyperparameters must be present: c('",
                paste0(collapse = "', '",  names(kernel$hyperparams))),
         "')")
  }
  hyperparams[names(kernel$hyperparams)]
}

File: src/kernel.cpp
#include "kernel.h"
using namespace Rcpp;

//TODO
kernel::kernel(int inputDimension): kernel(inputDimension, arma::vec({}),
               [](const arma::vec&, const arma::vec&, const arma::vec&) -> double {
                 return 0;
               }, arma::vec({}), arma::vec({}), 0) {
}

kernel::kernel(int inputDimension, const arma::vec& hyperparams,
       const std::function<double(const arma::vec&, const arma::vec&,const arma::vec&)>& expression,
       const arma::vec& lowerBound, const arma::vec& upperBound, double epsilon)
       // note the use of the , operator to sanitize the inputs
       : mInputDimension( (kernel::sanitize_input_dimension(inputDimension), inputDimension) ),
         mHP( (kernel::sanitize_hyperparams(hyperparams, lowerBound, upperBound),
               hyperparams) ),
         mExpression(expression),
         mLowerBound(lowerBound),
         mUpperBound(upperBound),
         mEpsilon( (kernel::check_epsilon(epsilon), epsilon) ){
  }


kernel::kernel(int inputDimension, const arma::vec& hyperparams,
               const std::function<double(const arma::vec&, const arma::vec&, const arma::vec&)>& expression,
               double epsilon)
  : mInputDimension( (kernel::sanitize_input_dimension(inputDimension), inputDimension)),
    mHP(hyperparams), mLowerBound(hyperparams.n_elem), mUpperBound(hyperparams.n_elem),
    mExpression(expression),
    mEpsilon( (kernel::check_epsilon(epsilon), epsilon) ) {
  mLowerBound.fill(0);
  mUpperBound.fill(std::numeric_limits<double>::max());
  check_hyperparameters_within_bounds(mHP, mLowerBound, mUpperBound);
}


kernel::~kernel() {
}

int kernel::get_input_dimension() const {
  return mInputDimension;
}

void kernel::set_hyperparams(const arma::vec& hyperparams) {
  if ((hyperparams.n_elem != mHP.n_elem)) {
    throw std::invalid_argument("Invalid hyperparams' length");
  }
  if (arma::any(hyperparams > mUpperBound) ||
      arma::any(hyperparams < mLowerBound)) {
    throw std::invalid_argument("hyperparams is not contained withing \
                                [lowerBound, upperBound");
  }
  mHP = hyperparams;
}


arma::vec kernel::get_hyperparams() const {
  return mHP;
}

// the lower bound only can be increased with respect to the initial value to
// ensure correctness
void kernel::increase_lower_bound(arma::vec& lowerBound){
  check_bounds_consistency(lowerBound, mUpperBound);
  if (arma::any(lowerBound < mLowerBound)) {
    throw std::invalid_argument("The lower bound can only be increased");
  }
  mLowerBound = lowerBound;
}


arma::vec kernel::get_lower_bound() const {
  return mLowerBound;
}


// the upper bound only can be decreased with respect to the initial value to
// ensure correctness
void kernel::decrease_upper_bound (arma::vec& upperBound){
  check_bounds_consistency(mLowerBound, upperBound);
  if (arma::any(upperBound > mUpperBound)) {
    throw std::invalid_argument("The upper bound can only be decreased");
  }
  mUpperBound = upperBound;
}


arma::vec kernel::get_upper_bound() const {
  return mUpperBound;
}

double kernel::get_epsilon() const {
  return mEpsilon;
}

void kernel::set_epsilon(double epsilon) {
  check_epsilon(epsilon);
  mEpsilon = epsilon;
}


arma::mat kernel::covmat(const arma::mat& x,
                         const arma::mat& y) const {
  int N = x.n_rows;
  int M = y.n_rows;
  int D = x.n_cols;

  if (D != y.n_cols) {
    throw std::invalid_argument("The dimensions of the vectors do not match");
  }

  check_input_dimension(D);

  arma::mat cov(N,M);
  arma::vec xi, xj;
  for (int i = 0; i < N; i++) {
    xi = arma::conv_to< arma::vec >::from(x.row(i));
    for (int j = 0; j < M; j++) {
      xj = arma::conv_to< arma::vec >::from(y.row(j));
      cov(i,j) = mExpression(xi, xj, mHP);
    }
  }
  // TODO: epsilon with covmat?
  return cov;
}

arma::mat kernel::autocovmat(const arma::mat& x) const {
  int N = x.n_rows;
  check_input_dimension(x.n_cols);

  arma::mat cov(N,N);
  arma::vec xi, xj;
  for (int i = 0; i < N; i++) {
    xi = arma::conv_to< arma::vec >::from(x.row(i));
    cov(i,i) = mExpression(xi,xi, mHP);
    for (int j = i + 1; j < N; j++) {
      xj = arma::conv_to< arma::vec >::from(x.row(j));
      cov(i,j) = mExpression(xi, xj, mHP);
      cov(j,i) = cov(i,j);
    }
  }
  // Add small fact or for better numerical stability
  if (mEpsilon != 0) {
    for (int i = 0; i < N; i++) {
      cov(i,i) += mEpsilon;
    }
  }
  return cov;
}

arma::vec kernel::variances(const arma::mat& x) const {
  int N = x.n_rows;

  check_input_dimension(x.n_cols);

  arma::colvec cov(N);
  arma::vec xi;
  for (int i = 0; i < N; i++) {
    xi = arma::conv_to< arma::vec >::from(x.row(i));
    cov(i) = mExpression(xi, xi, mHP);
  }
  // Add small fact or for consistency with autocov
  if (mEpsilon != 0) {
    for (int i = 0; i < N; i++) {
      cov(i) += mEpsilon;
    }
  }
  return cov;

}

kernel kernel::sum_kernel(const kernel& ker) {
  if (this->mInputDimension != ker.mInputDimension) {
    throw std::invalid_argument("The input Dimension of the Kernels doest not \
                                 match");
  }
  auto f1 = this->mExpression, f2 = ker.mExpression;
  int l1 = this->mHP.n_elem;
  int l2 = (ker.mHP).n_elem;
  std::function<double(const arma::vec&, const arma::vec&,
                       const arma::vec&)> exp =
                         [f1, f2, l1, l2](const arma::vec& x, const arma::vec& y,
                                          const arma::vec& hp) -> double {
                                            arma::vec hp1, hp2;
                                            if (l1 > 0) {
                                              hp1 =  hp.subvec(0, l1 - 1);
                                            }
                                            if (l2 > 0) {
                                              hp2 =  hp.subvec(l1, l1 + l2 - 1);
                                            }
                                            return f1(x, y, hp1) + f2(x, y, hp2);
                                          };
  return kernel(this->mInputDimension,
                arma::join_cols(this->mHP, ker.mHP), exp,
                arma::join_cols(this->mLowerBound, ker.mLowerBound),
                arma::join_cols(this->mUpperBound, ker.mUpperBound),
                this->mEpsilon + ker.mEpsilon);
}


kernel kernel::multiply_kernel(const kernel& ker) {
  if (this->mInputDimension != ker.mInputDimension) {
    throw std::invalid_argument("The input Dimension of the Kernels doest not \
                                  match");
  }
  auto f1 = this->mExpression, f2 = ker.mExpression;
  int l1 = this->mHP.n_elem;
  int l2 = (ker.mHP).n_elem;
  std::function<double(const arma::vec&, const arma::vec&,
                       const arma::vec&)> exp =
                         [f1, f2, l1, l2](const arma::vec& x, const arma::vec& y,
                                          const arma::vec& hp) -> double {
                                            arma::vec hp1, hp2;
                                            if (l1 > 0) {
                                              hp1 =  hp.subvec(0, l1 - 1);
                                            }
                                            if (l2 > 0) {
                                              hp2 =  hp.subvec(l1, l1 + l2 - 1);
                                            }
                                            return f1(x, y, hp1) * f2(x, y, hp2);
                                          };
  return kernel(this->mInputDimension,
                arma::join_cols(this->mHP, ker.mHP), exp,
                arma::join_cols(this->mLowerBound, ker.mLowerBound),
                arma::join_cols(this->mUpperBound, ker.mUpperBound),
                this->mEpsilon + ker.mEpsilon);
}

// constant * kernel
kernel kernel::scale_kernel(double constant) {
  if (constant < 0) {
    throw std::invalid_argument("Constant should be >= 0");
  }
  auto f1 = this->mExpression;
  int l1 = this->mHP.n_elem;
  std::function<double(const arma::vec&, const arma::vec&,
                       const arma::vec&)> exp =
                         [f1, constant](const arma::vec& x, const arma::vec& y,
                              const arma::vec& hp) -> double {
                                            return constant * f1(x, y, hp);
                              };
  return kernel(this->mInputDimension,
                this->mHP, exp,
                this->mLowerBound,
                this->mUpperBound,
                this->mEpsilon);
}


void kernel::sanitize_input_dimension(int vectorDimension) {
  if (vectorDimension < 1) {
    throw std::invalid_argument("Input dimension is smaller than 1");
  }
}

void kernel::sanitize_hyperparams(const arma::vec& hyperparams,
                                  const arma::vec& lowerBound,
                                  const arma::vec& upperBound) {
  if (lowerBound.n_elem != hyperparams.n_elem) {
    throw std::invalid_argument("Size of lower bound differs from size of \
                                 hyperparameters");
  }
  if (upperBound.n_elem != hyperparams.n_elem) {
    throw std::invalid_argument("Size of upper bound differs from size of \
                                 hyperparameters");
  }
  check_bounds_consistency(lowerBound, upperBound);
  check_hyperparameters_within_bounds(hyperparams, lowerBound, upperBound);
}


void kernel::check_hyperparameters_within_bounds(const arma::vec& hyperparams,
                                                 const arma::vec& lowerBound,
                                                 const arma::vec& upperBound) {
  if (arma::any(hyperparams > upperBound) || (arma::any(hyperparams < lowerBound))) {
    throw std::invalid_argument("The hyperparameters are not contained within \
                                 [lowerBound, upperBound]");
  }
}


void kernel::check_bounds_consistency(const arma::vec& lowerBound,
                                      const arma::vec& upperBound) {
  if (arma::any(lowerBound > upperBound)) {
    throw std::invalid_argument("Some values of the upper bound are \
                                 below the lower bound");
  }
}


void kernel::check_epsilon(double epsilon) {
  if (epsilon < 0) {
    throw std::invalid_argument("Epsilon should be >= 0");
  }
}

void kernel::check_input_dimension(int vectorDimension) const {
  if (vectorDimension != mInputDimension) {
    throw std::invalid_argument("The input Dimension of the Kernel doest not \
                                   match the dimension of the vectors");
  }
}
